packages/sourcecred/src/core/ledger/ledger.js

```diff
@@ -819,23 +920,99 @@ export class Ledger {
     if (this._grainIntegrationStatuses.get(id) === true) {
       throw new Error("Integration has already executed this distribution");
     }
-    this._createAndProcessEvent({
-      type: "MARK_DISTRIBUTION_EXECUTED",
-      id,
-    });
+    this._grainIntegrationStatuses.set(id, true);
+  }
+
+  /**
+   *   Setting a currency is a prerequisite to executing a grain integration.
+   *
+   *   When accounting is enabled, this function will set an external currency
+   *   in the ledger, but will not modify any accounts.
+   *
+   *   When accounting is disabled, this function can be called to update
+   *   the ledger's external currency configuration. Note that this will also
+   *   deactivate identities missing a payout address that matches the new
+   *   configuration when accounting is disabled.
+   */
+  setExternalCurrency(chainId: ChainId, tokenAddress?: EthAddress): Ledger {
+    const currency = buildCurrency(chainId, tokenAddress);
+
+    if (!deepEqual(currency, this._externalCurrency)) {
+      this._createAndProcessEvent({
+        type: "SET_EXTERNAL_CURRENCY",
+        currency,
+      });
+      // Deactivate accounts if they don't have an address stored for
+      // the newly configured currency.
+      if (!this._balanceAccountingEnabled) {
+        this._deactivateAccountsWithoutPayoutAddress();
+      }
```

Comment:
This should go in the private method under the mutations line, right?

Reply:
it can't. Calls to public methods from private methods result in out-of-order timestamps.

This is an acceptable location because it only calls public public methods under the hood. I'd like to keep this as a private method even though it only invokes a public method though because I don't want it to be a documented part of the API right now.

Reply:
Ah cool